src/rsa.rs

```diff
@@ -475,6 +486,51 @@ extern {
                                  rng: *mut rand::RAND) -> c::int;
 }
 
+#[cfg(feature = "rsa_signing")]
+#[allow(non_snake_case)]
+#[doc(hidden)]
+#[no_mangle]
+pub unsafe extern fn BN_rand_range(dest: *mut BIGNUM, range: *const BIGNUM,
+                                   rng: *mut rand::RAND) -> c::int {
```

Comment:
> (A) Instead of converting bytes to limbs each time we generate a number, we fill the limbs buffer directly.

Right.

> However, it's not immediately obvious to me that even so we'll be able to avoid writing new code to handle the case of range bounds which use only a few bits of the most significant limb, which motivated most of the code that I added in limb.rs.

If I understand the problem correctl : We could generate `num_limbs * BN_BYTES` of random bytes to fill the buffer. Then, in the highest limb, we need to mask `n` bits, where `n = (BN_num_bits(x) % BN_BITS2)` (or so). Then we do the comparison using GFp_constant_time_limbs_lt_limbs and try again if that fails. The issue is that it is a little bit unclear how to do the masking of the bits? Is that right?

Given n, `(1 << n) - 1` should give you the mask, I think. For example, we want to preserve 4 bits. `1 << 4 == 0b10000` and so `(1 << 4) - 1 == 0b1111`.

You can also look at the boringssl code at https://github.com/google/boringssl/blob/bfefc27c2bb4af62e09569e36b018d60da98a680/crypto/bn/random.c#L247 for some inspiration, in case my math above is off.